test/utils/integration/integration.go

```diff
@@ -0,0 +1,233 @@
+		Context:          ctx,
+		k8sClient:        k8sClient,
+		K8sDynamicClient: dynamicClient,
+		Reconciler:       eventingReconciler,
+		Logger:           sugaredLogger,
+		Recorder:         &recorder,
+		EnvTestInstance:  testEnv,
+		TestCancelFn:     cancelCtx,
+	}, nil
+}
+
+func (env TestEnvironment) TearDown() error {
+	if env.TestCancelFn != nil {
+		env.TestCancelFn()
+	}
+
+	// retry to stop the api-server
+	sleepTime := 1 * time.Second
+	var err error
+	const retries = 20
+	for i := 0; i < retries; i++ {
+		if err = env.EnvTestInstance.Stop(); err == nil {
+			break
+		}
+		time.Sleep(sleepTime)
+	}
+	return err
+}
+
+func StartEnvTest(projectRootDir string, celValidationEnabled bool) (*envtest.Environment, *rest.Config, error) {
+	// Reference: https://book.kubebuilder.io/reference/envtest.html
+	useExistingCluster := useExistingCluster
+	testEnv := &envtest.Environment{
+		CRDDirectoryPaths: []string{
+			filepath.Join(projectRootDir, "config", "crd", "bases"),
+		},
+		ErrorIfCRDPathMissing:    true,
+		AttachControlPlaneOutput: attachControlPlaneOutput,
+		UseExistingCluster:       &useExistingCluster,
+	}
+
+	args := testEnv.ControlPlane.GetAPIServer().Configure()
+	if celValidationEnabled {
+		args.Set("feature-gates", "CustomResourceValidationExpressions=true")
+	} else {
+		args.Set("feature-gates", "CustomResourceValidationExpressions=false")
+	}
+
+	var cfg *rest.Config
+	err := retry.Do(func() error {
+		defer func() {
+			if r := recover(); r != nil {
+				log.Println("panic recovered:", r)
+			}
+		}()
+		cfgLocal, startErr := testEnv.Start()
+		cfg = cfgLocal
+		return startErr
+	},
+		retry.Delay(testEnvStartDelay),
+		retry.DelayType(retry.FixedDelay),
+		retry.Attempts(testEnvStartAttempts),
+		retry.OnRetry(func(n uint, err error) {
+			log.Printf("[%v] try failed to start testenv: %s", n, err)
+			if stopErr := testEnv.Stop(); stopErr != nil {
+				log.Printf("failed to stop testenv: %s", stopErr)
+			}
+		}),
+	)
+	return testEnv, cfg, err
+}
+
+// GetEventingAssert fetches an Eventing from k8s and allows making assertions on it.
+func (env TestEnvironment) GetEventingAssert(g *gomega.GomegaWithT,
+	eventing *eventingv1alpha1.Eventing) gomega.AsyncAssertion {
+	return g.Eventually(func() *eventingv1alpha1.Eventing {
+		gotEventing, err := env.GetEventingFromK8s(eventing.Name, eventing.Namespace)
+		if err != nil {
+			log.Printf("fetch subscription %s/%s failed: %v", eventing.Name, eventing.Namespace, err)
+			return nil
+		}
+		return &gotEventing
+	}, BigTimeOut, SmallPollingInterval)
+}
+
+func (env TestEnvironment) GetEventingFromK8s(name, namespace string) (eventingv1alpha1.Eventing, error) {
+	var eventing eventingv1alpha1.Eventing
+	err := env.k8sClient.Get(env.Context, k8stypes.NamespacedName{
+		Name:      name,
+		Namespace: namespace,
+	}, &eventing)
+	return eventing, err
+}
+
+func (env TestEnvironment) EnsureNamespaceCreation(t *testing.T, namespace string) {
+	if namespace == "default" {
+		return
+	}
```

Comment:
(nit) This can be removed, thanks to `IgnoreAlreadyExists` below.

Reply:
Done